### DISCUSSION TITLE
I've just released `logot` - a log capture and assertion plugin 🚀

### DISCUSSION THREAD

**Discussion by etianen** [pytest-dev/pytest#11966] [Show and tell]
Hello! 👋 I've just released [logot](https://github.com/etianen/logot), which as the title says, is a log capture and assertion plugin. 

`logot` has a few unique things, such as being [logging-framework-agnostic](https://logot.readthedocs.io/latest/integrations/index.html#logging-frameworks) and having support for testing highly concurrent code using [threads](https://logot.readthedocs.io/latest/index.html#testing-threaded-code) or [async](https://logot.readthedocs.io/latest/index.html#testing-asynchronous-code). But those things are relatively niche. What I'd like to show here are a few examples of how it can be a nice `caplog` replacement, even in "normal" synchronous code.

## As a `caplog` replacement

Here's a really simple example testing that a piece of code logs as expected:

``` py
from logot import Logot, logged

def test_something(logot: Logot) -> None:
    do_something()
    logot.assert_logged(logged.info("Something was done"))
```

You'll see a couple of things here. A `logot` fixture, and a `logged` API. Use these together to make neat little log assertions. The equivalent code using `caplog` would be:

``` py
def test_something(caplog: pytest.LogCaptureFixture) -> None:
    do_something()
    assert any(
        record.levelno == logging.INFO and record.message == "Something was done"
        for record in caplog.records
    )
```

I think the `logot` code is clearer, and hopefully you do too! 🤗 

## Log message matching

One of `logot`s more useful features is the ability to match log messages using [`%`-style placeholders](https://logot.readthedocs.io/latest/log-message-matching.html) rather than regex. This syntax was chosen to be as close as possible to the `%` placeholders used by the stdlib `logging` library.

``` py
from logot import Logot, logged

def test_something(logot: Logot) -> None:
    do_something()
    # Match a string placeholder with `%s`.
    logot.assert_logged(logged.info("Something %s done"))
```

The equivalent using `caplog` gets pretty verbose:

``` py
import re

def test_something(caplog: pytest.LogCaptureFixture) -> None:
    do_something()
    assert any(
        record.levelno == logging.INFO and re.fullmatch("Something .*? done", record.message, re.DOTALL)
        for record in caplog.records
    )
```

If your message contained everyday punctuation like `.`, you have to start worrying about regex escaping too! I hope that `%`-style message matching gives a clearer, more _loggy_ way of matching log messages.

## Log pattern matching

This feature is generally aimed towards testing code using [threads](https://logot.readthedocs.io/latest/index.html#testing-threaded-code) or [async](https://logot.readthedocs.io/latest/index.html#testing-asynchronous-code), where messages can arrive out-of-order. But it's also useful for testing synchronous code.

``` py
from logot import Logot, logged

def test_app(logot: Logot) -> None:
   do_something()
   logot.wait_for(logged.info("Something happened") | logged.error("Something broke!"))
```

This example tests whether the `INFO` log "Something happened" **or** the `ERROR` "Something broke!" was emitted, and passes on either. The equivalent using `caplog` gets quite long:

``` py
def test_something(caplog: pytest.LogCaptureFixture) -> None:
    do_something()
    assert any(
        (record.levelno == logging.INFO and record.message == "Something happened")
        or (record.levelno == logging.ERROR and record.message == "Something broke!")
        for record in caplog.records
    )
```

## I hope you like it! ❤️ 

This is only a v1 release, but it's building on a lot of ideas I've been developing in different projects for a while now. I hope you like it, and find it useful.

The project [documentation](https://logot.readthedocs.io/latest/index.html) and [pytest integeration guide](https://logot.readthedocs.io/latest/using-unittest.html) are there if you'd like to find out more. 🙇

**Comment by caplogebenezere** [comment]
Caplog
